core/src/poh_timing_reporter.rs

```diff
@@ -0,0 +1,184 @@
+use {
+    solana_metrics::{datapoint::PohTimingPoint, datapoint_info},
+    solana_sdk::clock::Slot,
+    std::{collections::HashMap, fmt},
+};
+
+/// A PohTimestamp records timing of the events during the processing of a slot
+/// by the validator
+#[derive(Debug, Clone, Copy, Default)]
+pub struct SlotPohTimestamp {
+    /// Slot start time from poh
+    pub start_time: u64,
+    /// Slot end time from poh
+    pub end_time: u64,
+    /// Last shred received time from block producer
+    pub full_time: u64,
+}
+
+/// Display trait
+impl fmt::Display for SlotPohTimestamp {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "SlotPohTimestamp: start={} end={} full={}",
+            self.start_time, self.end_time, self.full_time
+        )
+    }
+}
+
+impl SlotPohTimestamp {
+    /// Return true if the timing points of all events are received.
+    pub fn is_complete(&self) -> bool {
+        self.start_time != 0 && self.end_time != 0 && self.full_time != 0
+    }
+
+    /// Update with timing point
+    pub fn update(&mut self, timing_point: PohTimingPoint) {
+        match timing_point {
+            PohTimingPoint::PohSlotStart(ts) => self.start_time = ts,
+            PohTimingPoint::PohSlotEnd(ts) => self.end_time = ts,
+            PohTimingPoint::FullSlotReceived(ts) => self.full_time = ts,
+        }
+    }
+
+    /// Return the time difference from slot full to slot start
+    pub fn slot_full_to_start_time(&self) -> i64 {
+        (self.start_time as i64).saturating_sub(self.full_time as i64)
+    }
+
+    /// Return the time difference from slot full to slot end
+    pub fn slot_full_to_end_time(&self) -> i64 {
+        (self.end_time as i64).saturating_sub(self.full_time as i64)
```

Comment:
Same as the comments for `slot_full_to_start_time` above, I think right now the likelihood is that slot full happens before slot end. Hence I think it makes sense to switch the subtraction around.

This is because given that leaders are allocated  4 blocks each, most of the time when transitioning between a parent slot `A` and child slot `B`, `A` and `B` are both from the same leader.

As the receiver of these blocks, we have to:
1. wait for at least one network hop `N` to get the last shreds from the leader for slot `A`
2. Finish replaying `A`
3. Then we reset Poh to slot `A` after finishing replaying slot `A`. 

However the leader doesn't have to eat the network latency for `N` and replaying the slot, he's already started the Poh for slot `B`, so he's running ahead. Hence his block `B` should arrive before we've completed our Poh for `B`

Reply:
I think the code does what you described above slot_end > slot_full. Right?

Reply:
ah yup brain fart, thanks